### ISSUE TITLE
Crazy idea #1: adding support for MongoDB

### ISSUE THREAD

**Issue by JohnAD** [moigagoo/norm#13]
A type object would be the perfect way for a MongoDB document database to store and retrieve objects.

Essentially, adding optional support for the `nimongo` library in addition to the `sqlite` and `postgres` databases.

A side effect is that the `postgres` will also start supporting:

  arrays - https://www.postgresql.org/docs/9.5/arrays.html
  composites - https://www.postgresql.org/docs/9.5/rowtypes.html

since MongoDB pretty much requires that also. I'm thinking composites would be done with references to other type objects. Such as:

``` nim
composite:
  type
    Address = object
      Street: string
      City: string
      State: string

db("petshop.db", "", "", ""):
  type       
    User = object
      age: int
      name: string
      home: Address
```

I'd be happy to write it and do a PR. But wanted to see if you were interested in it for the library.

**Comment by moigagoo** [comment]
I think it's a great idea! However, I have to warn you that I won't be able to assist you with it very much. I've never used arrays or composite types in PostgreSQL, and it's been a few years since I last used MongoDB (it even that was a rudimentary experience). Probably, the only thing I could assist you with is code review.

If you're ready to take it all on yourself—then feel free to. It'll be my pleasure to add this feature to norm, I think it would be a great addition.

**Comment by JohnAD** [comment]
I might do this. I'm about 50% convinced I might make my next web site in Nim. If I do so, then I'll do both this and the other crazy idea. :)  I'll let you know my decision in about a week. It sounds like a fun project.

**Comment by JohnAD** [comment]
I have decided that, yes, I will do this. In next few days, I will send a pull request on a new branch with "[WIP]" in the title so that you can see where I am in the progress of that branch.

**Comment by LantosTG** [comment]
how far did you get into testing this crazy idea @JohnAD ?

**Comment by JohnAD** [comment]
I actually gotten very very far. I've got a website that uses it and that website is near production release.

I've avoided requesting a merge because it is dependent on two new libraries that were forked from the source `nimongo` driver library and don't exist on **nimble** yet. The new mongo driver is very fundamentally different in how it handles threading, so a simple merge would not have worked without breaking other folks code.

So, your message is a good reminder; I need to finish this up. I will try to get them up this week. The new libraries, if curious, are at:

https://github.com/JohnAD/mongopool
https://github.com/JohnAD/bson

Once these are up, I will do some final cleanup; update the documentation, and send a good final PR to `norm` and ask for review.
